orca-spring-batch/src/main/java/com/netflix/spinnaker/orca/batch/stages/RollingPushStageDefinitionBuilder.java

```diff
@@ -0,0 +1,105 @@
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import com.netflix.spinnaker.orca.pipeline.model.Task;
+import org.springframework.batch.core.Step;
+import org.springframework.batch.core.job.builder.FlowBuilder;
+import static com.netflix.spinnaker.orca.ExecutionStatus.NOT_STARTED;
+import static com.netflix.spinnaker.orca.ExecutionStatus.REDIRECT;
+
+@Deprecated
+public class RollingPushStageDefinitionBuilder extends StageBuilder {
+
+  private final StageDefinitionBuilder delegate;
+  private final ExecutionListenerProvider executionListenerProvider;
+
+  public RollingPushStageDefinitionBuilder(StageDefinitionBuilder delegate,
+                                           ExecutionListenerProvider executionListenerProvider) {
+    super(delegate.getType());
+    this.delegate = delegate;
+    this.executionListenerProvider = executionListenerProvider;
+  }
+
+  @Override
+  protected FlowBuilder buildInternal(FlowBuilder jobBuilder, Stage stage) {
+    TaskGraph graph = delegate.buildTaskGraph(stage);
+    graph.forEach(node -> {
+      if (node instanceof TaskDefinition) {
+        TaskDefinition taskDef = (TaskDefinition) node;
+        addStep(jobBuilder, stage, taskDef);
+      } else if (node instanceof TaskGraph) {
+        TaskGraph subGraph = (TaskGraph) node;
+        List<TaskNode> taskNodes = Lists.newArrayList(subGraph.iterator());
+
+        Step startOfCycle = addStep(jobBuilder, stage, (TaskDefinition) taskNodes.remove(0));
+
+        while (taskNodes.size() > 1) {
+          addStep(jobBuilder, stage, (TaskDefinition) taskNodes.remove(0));
+        }
+
+        Step endOfCycle = addStep(jobBuilder, stage, (TaskDefinition) taskNodes.remove(0), new RedirectResetListener());
+        jobBuilder.on(REDIRECT.name()).to(startOfCycle);
+        jobBuilder.from(endOfCycle);
+      }
+    });
+    return jobBuilder;
+  }
+
+  private Step addStep(FlowBuilder jobBuilder, Stage stage, TaskDefinition taskDef) {
+    Step step = buildStep(stage, taskDef.getName(), taskDef.getImplementingClass());
+    jobBuilder.next(step);
+    return step;
+  }
+
+  private Step addStep(FlowBuilder jobBuilder, Stage stage, TaskDefinition taskDef, StageListener listener) {
+    Step step = buildStep(stage, taskDef.getName(), taskDef.getImplementingClass(), executionListenerProvider.wrap(listener));
+    jobBuilder.next(step);
+    return step;
+  }
+
+  /**
+   * A listener that resets the task status of everything in a loop so that it can be re-run without interfering
+   * with restart semantics.
+   */
+  static class RedirectResetListener implements StageListener {
+    @Override
+    public <T extends Execution<T>> void afterTask(Persister persister, Stage<T> stage, Task task, ExecutionStatus executionStatus, boolean wasSuccessful) {
+      if (executionStatus == REDIRECT) {
+        // TODO: same hideous index reference was in v1
+        stage.getTasks().subList(1, 11).forEach(it -> {
```

Comment:
https://github.com/spinnaker/orca/blob/master/orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/kato/pipeline/RollingPushStage.groovy#L103

Indexes are different because the old one is an inclusive index.